nibabel/brainvoyager/bv.py

```diff
@@ -0,0 +1,957 @@
+        ----------
+        binaryblock : {None, string} optional
+            binary block to set into header.  By default, None, in
+            which case we insert the default empty header block
+        endianness : {None, '<','>', other endian code} string, optional
+            endianness of the binaryblock.  If None, guess endianness
+            from the data.
+        check : bool, optional
+            Whether to check content of header in initialization.
+            Default is True.
+        offset : int, optional
+            offset of the actual data into to binary file (in bytes)
+        """
+        if endianness != self.default_endianness:
+            raise BvError('BV files are always little-endian')
+        self.endianness = self.default_endianness
+        if hdr_dict is None:
+            hdr_dict = _proto2default(self.hdr_dict_proto)
+        self._hdr_dict = hdr_dict
+        if offset is None:
+            self.set_data_offset(calc_BV_header_size(
+                self.hdr_dict_proto, self._hdr_dict))
+        if 'framing_cube' in self._hdr_dict:
+            self._framing_cube = self._hdr_dict['framing_cube']
+        else:
+            self._framing_cube = self._guess_framing_cube()
+        if check:
+            self.check_fix()
+        return
+
+    @classmethod
+    def from_fileobj(klass, fileobj, endianness=default_endianness,
+                     check=True):
+        """Return read structure with given or guessed endiancode.
+
+        Parameters
+        ----------
+        fileobj : file-like object
+           Needs to implement ``read`` method
+        endianness : None or endian code, optional
+           Code specifying endianness of read data
+
+        Returns
+        -------
+        header : BvFileHeader object
+           BvFileHeader object initialized from data in fileobj
+        """
+        hdr_dict = parse_BV_header(klass.hdr_dict_proto, fileobj)
+        offset = fileobj.tell()
+        return klass(hdr_dict, endianness, check, offset)
+
+    @classmethod
+    def from_header(klass, header=None, check=False):
+        """Class method to create header from another header.
+
+        Parameters
+        ----------
+        header : ``Header`` instance or mapping
+           a header of this class, or another class of header for
+           conversion to this type
+        check : {True, False}
+           whether to check header for integrity
+
+        Returns
+        -------
+        hdr : header instance
+           fresh header instance of our own class
+        """
+        # own type, return copy
+        if type(header) == klass:
+            obj = header.copy()
+            if check:
+                obj.check_fix()
+            return obj
+        # not own type, make fresh header instance
+        obj = klass(check=check)
+        if header is None:
+            return obj
+        try:  # check if there is a specific conversion routine
+            mapping = header.as_bv_map()
```

Comment:
Sure.   Each header defines its preferred map generation method, with default `None`.  So:


Header.map_method_name == None
AnalyzeHeader.map_method_name = 'as_analyze_map'
BvFileHeader.map_method_name = 'as_bv_map'.

Then the `Header` class has the following method (moved from the AnalyzeHeader class):

```
    @classmethod
    def from_header(klass, header=None, check=True):
        ''' Class method to create header from another header

        Parameters
        ----------
        header : ``Header`` instance or mapping
           a header of this class, or another class of header for
           conversion to this type
        check : {True, False}
           whether to check header for integrity

        Returns
        -------
        hdr : header instance
           fresh header instance of our own class
        '''
        # own type, return copy
        if type(header) == klass:
            obj = header.copy()
            if check:
                obj.check_fix()
            return obj
        # not own type, make fresh header instance
        obj = klass(check=check)
        if header is None:
            return obj
        # Check for method that returns a mapping this header understands
        map_method = (None if self.map_method_name is None
                                  else getattr(self.header, self.map_method_name, None))
        if map_method is not None:
            # header is convertible from a field mapping
            mapping = map_method()
            for key in mapping:
                try:
                    obj[key] = mapping[key]
                except (ValueError, KeyError):
                    # the presence of the mapping certifies the fields as being
                    # of the same meaning as for this class, so we can
                    # safely discard fields with names not known to this header
                    # type on the basis they are from the wrong dialect
                    pass
            # set any fields etc that are specific to this format (overriden by
            # sub-classes)
            obj._clean_after_mapping()
        # Fallback basic conversion always done.
        # More specific warning for unsupported datatypes
        orig_code = header.get_data_dtype()
        try:
            obj.set_data_dtype(orig_code)
        except HeaderDataError:
            raise HeaderDataError('Input header %s has datatype %s but '
                                  'output header %s does not support it'
                                  % (header.__class__,
                                     header.get_value_label('datatype'),
                                     klass))
        obj.set_data_dtype(header.get_data_dtype())
        obj.set_data_shape(header.get_data_shape())
        obj.set_zooms(header.get_zooms())
        if check:
            obj.check_fix()
        return obj
```

This the variant where the method goes in the Header class, but it could also be a mixin class, just containing this method, where AnalyzeHeader and BvFileHeader both inherit from both of Header and HeaderMapMixin, as in:

```
class BvFileHeader(Header, HeaderMapMixin):
```